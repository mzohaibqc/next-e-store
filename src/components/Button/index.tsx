'use client';

import classNames from 'classnames';

type Props = React.ComponentPropsWithoutRef<'button'> & {
  size?: 'small' | 'medium' | 'large';
};

export default function Button({
  className,
  size = 'medium',
  children,
  ...rest
}: Props) {
  return (
    <button
      className={classNames(
        'text-gray-900 dark:text-gray-200 hover:bg-gray-100 dark:hover:bg-gray-700 focus:outline-none focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-700 rounded-md text-sm p-1 inline-flex items-center justify-center',
        className,
        {
          'w-5 h-5': size === 'small',
          'w-6 h-6': size === 'medium',
          'w-8 h-8': size === 'large',
        }
      )}
      {...rest}
    >
      {children}
    </button>
  );
}
