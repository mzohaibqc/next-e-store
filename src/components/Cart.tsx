'use client';

import { useCart } from '@/store';

export default function Cart() {
  const { total } = useCart();
  return (
    <div className="w-full lg:w-[300px] lg:h-[400px] p-4 rounded-md border bg-white flex justify-between">
      <p className="text-lg font-medium">Total:</p>
      <p className="text-2xl font-medium">&pound; {total.toFixed(2)}</p>
    </div>
  );
}
