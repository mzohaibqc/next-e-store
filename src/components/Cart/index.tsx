'use client';

import { useCart } from '@/store';

export default function Cart() {
  const { total } = useCart();
  return (
    <div className="w-full lg:w-[300px] lg:h-[400px] p-4 rounded-md border bg-white flex justify-between">
      <p className="text-lg font-medium">Cart Total:</p>
      <p className="text-2xl font-medium">
        &pound;<span data-testid="cart-total">{total.toFixed(2)}</span>
      </p>
    </div>
  );
}
