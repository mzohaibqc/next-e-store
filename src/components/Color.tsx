type ColorProps = {
  color: string;
};

const colorMap: Record<string, string> = {
  Black: '#000000',
  Red: '#FF0000',
  Stone: '#EFCFB6',
};

export default function Color({ color }: ColorProps) {
  return (
    <div
      className="w-6 h-6 rounded-md bg-white shadow-[0_8px_30px_rgb(0,0,0,0.12)]"
      style={{ backgroundColor: colorMap[color] || color }}
    >
      <span className="sr-only color-value">{color}</span>
    </div>
  );
}
