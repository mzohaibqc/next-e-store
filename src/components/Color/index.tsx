import classNames from 'classnames';

type Props = React.ComponentPropsWithoutRef<'div'> & {
  color: string;
};

export default function Color({ color, ...rest }: Props) {
  return (
    <div
      {...rest}
      className={classNames(
        'w-6 h-6 rounded-md bg-white shadow-[0_8px_30px_rgb(0,0,0,0.12)]',
        rest.className
      )}
      style={{ backgroundColor: colorMap[color] || color }}
    >
      <span className="sr-only color-value">{color}</span>
    </div>
  );
}

export const colorMap: Record<string, string> = {
  Black: '#000000',
  Red: '#FF0000',
  Stone: '#EFCFB6',
};
