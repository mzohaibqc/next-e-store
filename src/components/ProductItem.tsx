'use client';
import { useCart } from '@/store';
import classNames from 'classnames';
import Image from 'next/image';
import { useCallback, useMemo } from 'react';
import Button from './Button';
import Color from './Color';
import { Add, Minus, Trash } from './Icons';

export type Product = {
  id: number;
  colour: string;
  name: string;
  price: number;
  img: string;
};

type Props = {
  product: Product;
  className?: string;
};

export default function ProductItem({ product, className }: Props) {
  const { addItem, removeItem, clearItem, cart } = useCart();
  const count = useMemo(
    () => cart.find((item) => item.id === product.id)?.count || 0,
    [cart, product]
  );
  const handleAddItem = useCallback(() => {
    addItem(product);
  }, [addItem, product]);
  const handleRemoveItem = useCallback(() => {
    removeItem(product);
  }, [removeItem, product]);
  const handleClearItem = useCallback(() => {
    clearItem(product);
  }, [clearItem, product]);
  const disabled = useMemo(() => count === 0, [count]);
  return (
    <div className={className}>
      <div className="flex flex-col sm:flex-row rounded-md overflow-hidden bg-white p-2 shadow-[0_8px_30px_rgb(0,0,0,0.12)]">
        <div className="img relative">
          <Image
            src={product.img}
            width={150}
            height={280}
            alt="Product Image"
            style={{ objectFit: 'contain', width: 'auto' }}
            className="rounded-md"
            priority
          />
        </div>
        <div className="info flex-1 px-4 py-2 flex flex-col">
          <div className="flex justify-between">
            <h2 className="text-xl font-semibold">{product.name}</h2>
            <Button
              onClick={handleClearItem}
              disabled={disabled}
              className={classNames('text-lg icon-btn')}
              aria-label="Remove Item"
            >
              <Trash />
            </Button>
          </div>
          <div className="flex items-center space-x-4 mt-10">
            <p className="text-lg font-medium">Price:</p>
            <p className="text-2xl font-medium">
              &pound;{' '}
              <span data-testid="product-price">
                {product.price.toFixed(2)}
              </span>
            </p>
          </div>
          <div className="flex items-center space-x-4">
            <p className="text-lg font-medium">Color:</p>
            <Color color={product.colour} />
          </div>
          <div className="flex-1"></div>
          <div className="flex flex-col justify-end items-end w-full mt-6">
            <div className="w-full border border-gray-200 px-2 py-1 rounded-md max-w-xs flex justify-around items-center">
              <Button
                onClick={handleRemoveItem}
                disabled={disabled}
                className={classNames('text-lg icon-btn')}
                aria-label="Remove one Item"
              >
                <Minus />
              </Button>
              <span className="text-lg px-5  border rounded-md text-center">
                {count}
              </span>
              <Button
                onClick={handleAddItem}
                className="text-lg icon-btn"
                aria-label="Add one Item"
              >
                <Add />
              </Button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
