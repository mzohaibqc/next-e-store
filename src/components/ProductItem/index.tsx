'use client';
import classNames from 'classnames';
import Image from 'next/image';
import { useCallback, useMemo } from 'react';
import Button from '@/components/Button';
import Color from '@/components/Color';
import { Add, Minus, Trash } from '@/components/Icons';
import { useCart } from '@/store';

export type Product = {
  id: number;
  colour: string;
  name: string;
  price: number;
  img: string;
};

type Props = {
  product: Product;
  className?: string;
};

export default function ProductItem({ product, className }: Props) {
  const { addItem, removeItem, clearItem, cart } = useCart();
  const count = useMemo(
    () => cart.find((item) => item.id === product.id)?.count || 0,
    [cart, product]
  );
  const handleAddItem = useCallback(() => addItem(product), [addItem, product]);
  const handleRemoveItem = useCallback(
    () => removeItem(product),
    [removeItem, product]
  );
  const handleClearItem = useCallback(
    () => clearItem(product),
    [clearItem, product]
  );
  const disabled = useMemo(() => count === 0, [count]);
  return (
    <div className={className} data-testid={`product-${product.id}`}>
      <div className="flex flex-col sm:flex-row rounded-md overflow-hidden bg-white p-2 shadow-[0_8px_30px_rgb(0,0,0,0.12)]">
        <div className="img relative w-full h-[380px] sm:w-[150px] sm:h-[220px]">
          <Image
            fill
            priority
            src={product.img}
            alt="Product Image"
            className="rounded-md"
            data-testid={`product-${product.id}-img`}
            style={{ objectFit: 'cover', objectPosition: '50% 30%' }}
          />
        </div>
        <div className="info flex-1 px-2 sm:px-4 py-2 flex flex-col">
          <div className="flex justify-between">
            <h2 className="text-xl font-semibold">{product.name}</h2>
            {/* <Button
              onClick={handleClearItem}
              disabled={disabled}
              className={classNames('text-lg icon-btn')}
              aria-label="Remove Item"
              data-testid={`product-${product.id}-clear`}
            >
              <Trash />
            </Button> */}
          </div>

          <div className="flex-1 flex flex-col items-center sm:items-end justify-end gap-3">
            <div className="flex justify-between items-center w-full sm:max-w-xs mt-6">
              <div className="flex items-center space-x-3">
                <p className="text-lg font-medium">Color:</p>
                <Color
                  color={product.colour}
                  data-testid={`product-${product.id}-color`}
                />
              </div>
              <div className="flex items-center space-x-3">
                <p className="text-lg font-medium">Price:</p>
                <p className="text-xl font-medium">
                  &pound;
                  <span data-testid={`product-${product.id}-price`}>
                    {product.price.toFixed(2)}
                  </span>
                </p>
              </div>
            </div>
            <div className="w-full border border-gray-200 px-2 py-1 rounded-md sm:max-w-xs flex justify-around items-center">
              <Button
                onClick={handleRemoveItem}
                disabled={disabled}
                className={classNames('text-lg icon-btn')}
                aria-label="Remove One Item"
                data-testid={`product-${product.id}-remove`}
              >
                <Minus />
              </Button>
              <span
                data-testid={`product-${product.id}-count`}
                className="text-lg px-5  border rounded-md text-center"
              >
                {count}
              </span>
              <Button
                onClick={handleAddItem}
                className="text-lg icon-btn"
                aria-label="Add One Item"
                data-testid={`product-${product.id}-add`}
              >
                <Add />
              </Button>
            </div>
            <button
              onClick={handleClearItem}
              disabled={disabled}
              data-testid={`product-${product.id}-clear`}
              className="w-full sm:max-w-xs text-white bg-gray-800 hover:bg-gray-900 focus:outline-none focus:ring-2 focus:ring-gray-300 font-medium rounded-md text-sm px-5 py-2.5 disabled:bg-gray-400 disabled:cursor-not-allowed dark:bg-gray-800 dark:hover:bg-gray-700 dark:focus:ring-gray-700 dark:border-gray-700"
            >
              Remove all Items
            </button>
          </div>
        </div>
      </div>
    </div>
  );
}
