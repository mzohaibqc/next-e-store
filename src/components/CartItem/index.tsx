'use client';
import classNames from 'classnames';
import Image from 'next/image';
import { useCallback, useMemo } from 'react';
import Button from '@/components/Button';
import Color from '@/components/Color';
import { Add, Minus } from '@/components/Icons';
import { useCart } from '@/store';
import { Product } from '../ProductItem';

type Props = {
  product: Product;
  className?: string;
};

export default function CartItem({ product, className }: Props) {
  const { addItem, removeItem, clearItem, cart } = useCart();
  const count = useMemo(
    () => Number(cart.find((item) => item.id === product.id)?.count),
    [cart, product]
  );
  const handleAddItem = useCallback(() => addItem(product), [addItem, product]);
  const handleRemoveItem = useCallback(
    () => removeItem(product),
    [removeItem, product]
  );
  const handleClearItem = useCallback(
    () => clearItem(product),
    [clearItem, product]
  );
  const disabled = useMemo(() => count === 0, [count]);
  return (
    <div
      className={classNames(className, 'flex flex-col')}
      data-testid={`product-${product.id}`}
    >
      <div className="flex-1 flex rounded-md overflow-hidden bg-white dark:bg-gray-900 shadow-[rgba(17,_17,_26,_0.1)_0px_0px_16px] border border-gray-200 dark:border-gray-800">
        <Image
          priority
          src={product.img}
          alt="Product Image"
          className="rounded-l-md"
          width={50}
          height={80}
          data-testid={`product-${product.id}-img`}
          style={{ objectFit: 'cover', objectPosition: '50% 20%' }}
        />
        <div className="info flex-1 px-2 py-1 flex flex-col">
          <h2
            className="text-sm font-semibold mb-1 line-clamp-1"
            title={product.name}
            aria-label={product.name}
          >
            {product.name}
          </h2>

          <div className="flex justify-between items-center gap-3 mb-1">
            <div className="flex items-center gap-3">
              <Color
                color={product.colour}
                data-testid={`product-${product.id}-color`}
                size="small"
              />
              <p className="text-sm text-gray-500 font-medium font-mono">
                &pound;
                <span data-testid={`product-${product.id}-price`}>
                  {product.price.toFixed(2)}
                </span>
              </p>
            </div>
            <button
              className="px-1 text-sm text-gray-600 disabled:text-gray-400 disabled:cursor-not-allowed disabled:hover:no-underline hover:underline focus:underline underline-offset-4 decoration-2"
              onClick={handleClearItem}
              disabled={disabled}
              data-testid={`product-${product.id}-clear`}
            >
              <span>Remove</span>
            </button>
          </div>
          <div className="w-full border border-gray-200 dark:border-gray-800 px-2 rounded-md sm:max-w-xs flex justify-around items-center">
            <Button
              onClick={handleRemoveItem}
              disabled={disabled}
              aria-label="Remove One Item"
              data-testid={`product-${product.id}-remove`}
              size="small"
            >
              <Minus className="w-4 h-4" />
            </Button>
            <span
              data-testid={`product-${product.id}-count`}
              className="text-md text-gray-500 px-5 min-w-16 rounded-md text-center"
            >
              {count}
            </span>
            <Button
              onClick={handleAddItem}
              aria-label="Add One Item"
              data-testid={`product-${product.id}-add`}
              size="small"
            >
              <Add className="w-4 h-4" />
            </Button>
          </div>
        </div>
      </div>
    </div>
  );
}
