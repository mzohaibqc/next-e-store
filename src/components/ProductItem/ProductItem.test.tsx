import { fireEvent, render, screen } from '@testing-library/react';
import { expect, describe, test, vi } from 'vitest';
import ProductItem from '@/components/ProductItem';
import { type Product } from '@/components/ProductItem';
import * as store from '@/store';

const product: Product = {
  id: 1,
  name: 'Product 1',
  price: 10,
  img: '/logo.png',
  colour: 'Black',
};
const cartItem: store.CartItem = {
  ...product,
  count: 10,
};

describe('ProductItem', () => {
  const useCartSpy = vi.spyOn(store, 'useCart');
  const mockCart = {
    total: 100,
    cart: [cartItem],
    addItem: vi.fn(),
    removeItem: vi.fn(),
    clearItem: vi.fn(),
  };
  useCartSpy.mockReturnValue(mockCart);
  test('ProductItem should show all data for given product', () => {
    const { container } = render(<ProductItem product={product} />);
    // Verify product name
    expect(screen.getByText(product.name)).toBeDefined;
    // Verify product price
    expect(
      screen.getByTestId(`product-${product.id}-price`).textContent
    ).toEqual(product.price.toFixed(2));
    // verify image src and alt text
    const image = screen.getByTestId(`product-${product.id}-img`);
    expect(decodeURIComponent(image.getAttribute('src') as string)).contain(
      product.img
    );
    expect(image.getAttribute('alt')).toEqual('Product Image');
    // Verify product colour
    expect(container.querySelector('.color-value')!.textContent).toEqual(
      product.colour
    );
    // Verify product count
    expect(
      screen.getByTestId(`product-${product.id}-count`).textContent
    ).toEqual(cartItem.count.toString());
    // Verify actions/buttons
    const addButton = screen.getByLabelText('Add One Item');
    expect(addButton).toBeDefined;
    fireEvent.click(addButton);
    expect(mockCart.addItem.mock.calls[0][0]).toEqual(product);
    const removeButton = screen.getByLabelText('Remove One Item');
    expect(removeButton).toBeDefined;
    fireEvent.click(removeButton);
    expect(mockCart.removeItem.mock.calls[0][0]).toEqual(product);
    const clearButton = screen.getByTestId(`product-${product.id}-clear`);
    expect(clearButton).toBeDefined;
    fireEvent.click(clearButton);
    expect(mockCart.clearItem.mock.calls[0][0]).toEqual(product);

    // Create snapshot
    expect(container).toMatchSnapshot();
  });
});
