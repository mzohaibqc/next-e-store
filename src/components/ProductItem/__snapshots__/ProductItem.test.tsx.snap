// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductItem > ProductItem should show all data for given product 1`] = `
<div>
  <div
    data-testid="product-1"
  >
    <div
      class="flex flex-col sm:flex-row rounded-md overflow-hidden bg-white p-2 shadow-[0_8px_30px_rgb(0,0,0,0.12)]"
    >
      <div
        class="img relative w-full h-[380px] sm:w-[150px] sm:h-[220px]"
      >
        <img
          alt="Product Image"
          class="rounded-md"
          data-nimg="fill"
          data-testid="product-1-img"
          decoding="async"
          fetchpriority="high"
          sizes="100vw"
          src="/_next/image?url=%2Flogo.png&w=3840&q=75"
          srcset="/_next/image?url=%2Flogo.png&w=640&q=75 640w, /_next/image?url=%2Flogo.png&w=750&q=75 750w, /_next/image?url=%2Flogo.png&w=828&q=75 828w, /_next/image?url=%2Flogo.png&w=1080&q=75 1080w, /_next/image?url=%2Flogo.png&w=1200&q=75 1200w, /_next/image?url=%2Flogo.png&w=1920&q=75 1920w, /_next/image?url=%2Flogo.png&w=2048&q=75 2048w, /_next/image?url=%2Flogo.png&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; object-position: 50% 30%; color: transparent;"
        />
      </div>
      <div
        class="info flex-1 px-2 sm:px-4 py-2 flex flex-col"
      >
        <div
          class="flex justify-between"
        >
          <h2
            class="text-xl font-semibold"
          >
            Product 1
          </h2>
        </div>
        <div
          class="flex-1 flex flex-col items-center sm:items-end justify-end gap-3"
        >
          <div
            class="flex justify-between items-center w-full sm:max-w-xs mt-6"
          >
            <div
              class="flex items-center space-x-3"
            >
              <p
                class="text-lg font-medium"
              >
                Color:
              </p>
              <div
                class="w-6 h-6 rounded-md bg-white shadow-[0_8px_30px_rgb(0,0,0,0.12)]"
                data-testid="product-1-color"
                style="background-color: rgb(0, 0, 0);"
              >
                <span
                  class="sr-only color-value"
                >
                  Black
                </span>
              </div>
            </div>
            <div
              class="flex items-center space-x-3"
            >
              <p
                class="text-lg font-medium"
              >
                Price:
              </p>
              <p
                class="text-xl font-medium"
              >
                Â£
                <span
                  data-testid="product-1-price"
                >
                  10.00
                </span>
              </p>
            </div>
          </div>
          <div
            class="w-full border border-gray-200 px-2 py-1 rounded-md sm:max-w-xs flex justify-around items-center"
          >
            <button
              aria-label="Remove One Item"
              class="w-8 h-8 rounded-md flex items-center justify-center cursor-pointer disabled:cursor-not-allowed text-lg hover:bg-gray-100 hover:border hover:border-gray-300 text-lg icon-btn"
              data-testid="product-1-remove"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                />
              </svg>
            </button>
            <span
              class="text-lg px-5  border rounded-md text-center"
              data-testid="product-1-count"
            >
              10
            </span>
            <button
              aria-label="Add One Item"
              class="w-8 h-8 rounded-md flex items-center justify-center cursor-pointer disabled:cursor-not-allowed text-lg hover:bg-gray-100 hover:border hover:border-gray-300 text-lg icon-btn"
              data-testid="product-1-add"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                />
              </svg>
            </button>
          </div>
          <button
            class="w-full sm:max-w-xs text-white bg-gray-800 hover:bg-gray-900 focus:outline-none focus:ring-2 focus:ring-gray-300 font-medium rounded-md text-sm px-5 py-2.5 disabled:bg-gray-400 disabled:cursor-not-allowed dark:bg-gray-800 dark:hover:bg-gray-700 dark:focus:ring-gray-700 dark:border-gray-700"
            data-testid="product-1-clear"
          >
            Remove all Items
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
